When pattern doesn't match the type of if let expression, see if a field would be of the correct type and provide an apprpriate structured suggestion

Just encountered the following output:

```
error[E0308]: mismatched types
   --> compiler/rustc_passes/src/loops.rs:152:40
    |
152 |   ...                   if let hir::ExprKind::Path(hir::QPath::Resolved(
    |  ______________________________^
153 | | ...                       None,
154 | | ...                       Path { ident, res: hir::def::Res::Err, .. },
155 | | ...                   )) = break_expr
    | |                        ^   ---------- this expression has type `&&rustc_hir::Expr<'_>`
    | |________________________|
    |                          expected struct `rustc_hir::Expr`, found enum `rustc_hir::ExprKind`
```

It would be nice if we could peek at the fields of `break_expr` and provide an appropriate structured suggestion:


```
error[E0308]: mismatched types
   --> compiler/rustc_passes/src/loops.rs:152:40
    |
152 |   ...                   if let hir::ExprKind::Path(hir::QPath::Resolved(
    |  ______________________________^
153 | | ...                       None,
154 | | ...                       Path { ident, res: hir::def::Res::Err, .. },
155 | | ...                   )) = break_expr
    | |                        ^   ---------- this expression has type `&&rustc_hir::Expr<'_>`
    | |________________________|
    |                          expected struct `rustc_hir::Expr`, found enum `rustc_hir::ExprKind`
    |
help: you might have meant to use field `kind` of type `rustc_hir::ExprKind`
    |
155 | ...                   )) = break_expr.kind
    |                                      ^^^^^
```

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"matsujika"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_infer/src/infer/error_reporting/mod.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/mod.rs
@@ -1838,9 +1838,6 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 .iter()
                 .filter(|field| field.vis.is_accessible_from(field.did, self.tcx))
                 .map(|field| (field.ident.name, field.ty(self.tcx, expected_substs)))
-                .inspect(|(name, ty)| {
-                    debug!("suggest_field_where_appropriate: name={:?}, ty={:?}", name, ty)
-                })
                 .find(|(_, ty)| ty::TyS::same_type(ty, exp_found.found))
             {
                 if let ObligationCauseCode::Pattern { span: Some(span), .. } = cause.code {

--- a/compiler/rustc_infer/src/infer/error_reporting/mod.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/mod.rs
@@ -1661,6 +1661,7 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         debug!("exp_found {:?} terr {:?}", exp_found, terr);
         if let Some(exp_found) = exp_found {
             self.suggest_as_ref_where_appropriate(span, &exp_found, diag);
+            self.suggest_field_where_appropriate(cause, &exp_found, diag);
             self.suggest_await_on_expect_found(cause, span, &exp_found, diag);
         }
 
@@ -1816,6 +1817,46 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 );
             }
             _ => {}
+        }
+    }
+
+    fn suggest_field_where_appropriate(
+        &self,
+        cause: &ObligationCause<'tcx>,
+        exp_found: &ty::error::ExpectedFound<Ty<'tcx>>,
+        diag: &mut DiagnosticBuilder<'tcx>,
+    ) {
+        debug!("suggest_field_where_appropriate(cause={:?}, exp_found={:?})", cause, exp_found);
+        if let ty::Adt(expected_def, expected_substs) = exp_found.expected.kind() {
+            if expected_def.is_enum() {
+                return;
+            }
+
+            if let Some((name, ty)) = expected_def
+                .non_enum_variant()
+                .fields
+                .iter()
+                .filter(|field| field.vis.is_accessible_from(field.did, self.tcx))
+                .map(|field| (field.ident.name, field.ty(self.tcx, expected_substs)))
+                .inspect(|(name, ty)| {
+                    debug!("suggest_field_where_appropriate: name={:?}, ty={:?}", name, ty)
+                })
+                .find(|(_, ty)| ty::TyS::same_type(ty, exp_found.found))
+            {
+                if let ObligationCauseCode::Pattern { span: Some(span), .. } = cause.code {
+                    if let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(span) {
+                        diag.span_suggestion(
+                            span,
+                            &format!(
+                                "you might have meant to use field `{}` of type `{}`",
+                                name, ty
+                            ),
+                            format!("{}.{}", snippet, name),
+                            Applicability::MaybeIncorrect,
+                        );
+                    }
+                }
+            }
         }
     }
 

--- a/compiler/rustc_infer/src/infer/error_reporting/mod.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/mod.rs
@@ -1661,7 +1661,7 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         debug!("exp_found {:?} terr {:?}", exp_found, terr);
         if let Some(exp_found) = exp_found {
             self.suggest_as_ref_where_appropriate(span, &exp_found, diag);
-            self.suggest_field_where_appropriate(cause, &exp_found, diag);
+            self.suggest_accessing_field_where_appropriate(cause, &exp_found, diag);
             self.suggest_await_on_expect_found(cause, span, &exp_found, diag);
         }
 
@@ -1820,7 +1820,7 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         }
     }
 
-    fn suggest_field_where_appropriate(
+    fn suggest_accessing_field_where_appropriate(
         &self,
         cause: &ObligationCause<'tcx>,
         exp_found: &ty::error::ExpectedFound<Ty<'tcx>>,

--- a/compiler/rustc_infer/src/infer/error_reporting/mod.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/mod.rs
@@ -1826,7 +1826,10 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         exp_found: &ty::error::ExpectedFound<Ty<'tcx>>,
         diag: &mut DiagnosticBuilder<'tcx>,
     ) {
-        debug!("suggest_accessing_field_where_appropriate(cause={:?}, exp_found={:?})", cause, exp_found);
+        debug!(
+            "suggest_accessing_field_where_appropriate(cause={:?}, exp_found={:?})",
+            cause, exp_found
+        );
         if let ty::Adt(expected_def, expected_substs) = exp_found.expected.kind() {
             if expected_def.is_enum() {
                 return;

--- a/compiler/rustc_infer/src/infer/error_reporting/mod.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/mod.rs
@@ -1842,13 +1842,20 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             {
                 if let ObligationCauseCode::Pattern { span: Some(span), .. } = cause.code {
                     if let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(span) {
+                        let suggestion = if expected_def.is_struct() {
+                            format!("{}.{}", snippet, name)
+                        } else if expected_def.is_union() {
+                            format!("unsafe {{ {}.{} }}", snippet, name)
+                        } else {
+                            return;
+                        };
                         diag.span_suggestion(
                             span,
                             &format!(
                                 "you might have meant to use field `{}` of type `{}`",
                                 name, ty
                             ),
-                            format!("{}.{}", snippet, name),
+                            suggestion,
                             Applicability::MaybeIncorrect,
                         );
                     }

--- a/compiler/rustc_infer/src/infer/error_reporting/mod.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/mod.rs
@@ -1826,7 +1826,7 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         exp_found: &ty::error::ExpectedFound<Ty<'tcx>>,
         diag: &mut DiagnosticBuilder<'tcx>,
     ) {
-        debug!("suggest_field_where_appropriate(cause={:?}, exp_found={:?})", cause, exp_found);
+        debug!("suggest_accessing_field_where_appropriate(cause={:?}, exp_found={:?})", cause, exp_found);
         if let ty::Adt(expected_def, expected_substs) = exp_found.expected.kind() {
             if expected_def.is_enum() {
                 return;

--- a/src/test/ui/suggestions/field-access.fixed
+++ b/src/test/ui/suggestions/field-access.fixed
@@ -12,15 +12,12 @@ enum B {
 
 fn main() {
     let a = A { b: B::Fst };
-    if let B::Fst = a.b {};
-    //~^ ERROR mismatched types [E0308]
-    // note: you might have meant to use field `b` of type `B`
+    if let B::Fst = a.b {}; //~ ERROR mismatched types [E0308]
+    //~^ HELP you might have meant to use field `b` of type `B`
     match a.b {
-        B::Fst => (),
-        B::Snd => (),
+        //~^ HELP you might have meant to use field `b` of type `B`
+        //~| HELP you might have meant to use field `b` of type `B`
+        B::Fst => (), //~ ERROR mismatched types [E0308]
+        B::Snd => (), //~ ERROR mismatched types [E0308]
     }
-    //~^^^ ERROR mismatched types [E0308]
-    // note: you might have meant to use field `b` of type `B`
-    //~^^^^ ERROR mismatched types [E0308]
-    // note: you might have meant to use field `b` of type `B`
 }

--- a/src/test/ui/suggestions/field-access.fixed
+++ b/src/test/ui/suggestions/field-access.fixed
@@ -10,6 +10,11 @@ enum B {
     Snd,
 }
 
+union Foo {
+    bar: u32,
+    qux: f32,
+}
+
 fn main() {
     let a = A { b: B::Fst };
     if let B::Fst = a.b {}; //~ ERROR mismatched types [E0308]
@@ -19,5 +24,12 @@ fn main() {
         //~| HELP you might have meant to use field `b` of type `B`
         B::Fst => (), //~ ERROR mismatched types [E0308]
         B::Snd => (), //~ ERROR mismatched types [E0308]
+    }
+
+    let foo = Foo { bar: 42 };
+    match unsafe { foo.bar } {
+        //~^ HELP you might have meant to use field `bar` of type `u32`
+        1u32 => (), //~ ERROR mismatched types [E0308]
+        _ => (),
     }
 }

--- a/src/test/ui/suggestions/field-access.fixed
+++ b/src/test/ui/suggestions/field-access.fixed
@@ -12,10 +12,10 @@ enum B {
 
 fn main() {
     let a = A { b: B::Fst };
-    if let B::Fst = a {};
+    if let B::Fst = a.b {};
     //~^ ERROR mismatched types [E0308]
     // note: you might have meant to use field `b` of type `B`
-    match a {
+    match a.b {
         B::Fst => (),
         B::Snd => (),
     }

--- a/src/test/ui/suggestions/field-access.rs
+++ b/src/test/ui/suggestions/field-access.rs
@@ -12,4 +12,12 @@ fn main() {
     if let B::Fst = a {};
     //~^ ERROR mismatched types [E0308]
     // note: you might have meant to use field `b` of type `B`
+    match a {
+        B::Fst => (),
+        B::Snd => (),
+    }
+    //~^^^ ERROR mismatched types [E0308]
+    // note: you might have meant to use field `b` of type `B`
+    //~^^^^ ERROR mismatched types [E0308]
+    // note: you might have meant to use field `b` of type `B`
 }

--- a/src/test/ui/suggestions/field-access.rs
+++ b/src/test/ui/suggestions/field-access.rs
@@ -1,0 +1,15 @@
+struct A {
+    b: B,
+}
+
+enum B {
+    Fst,
+    Snd,
+}
+
+fn main() {
+    let a = A { b: B::Fst };
+    if let B::Fst = a {};
+    //~^ ERROR mismatched types [E0308]
+    // note: you might have meant to use field `b` of type `B`
+}

--- a/src/test/ui/suggestions/field-access.rs
+++ b/src/test/ui/suggestions/field-access.rs
@@ -12,15 +12,12 @@ enum B {
 
 fn main() {
     let a = A { b: B::Fst };
-    if let B::Fst = a {};
-    //~^ ERROR mismatched types [E0308]
-    // note: you might have meant to use field `b` of type `B`
+    if let B::Fst = a {}; //~ ERROR mismatched types [E0308]
+    //~^ HELP you might have meant to use field `b` of type `B`
     match a {
-        B::Fst => (),
-        B::Snd => (),
+        //~^ HELP you might have meant to use field `b` of type `B`
+        //~| HELP you might have meant to use field `b` of type `B`
+        B::Fst => (), //~ ERROR mismatched types [E0308]
+        B::Snd => (), //~ ERROR mismatched types [E0308]
     }
-    //~^^^ ERROR mismatched types [E0308]
-    // note: you might have meant to use field `b` of type `B`
-    //~^^^^ ERROR mismatched types [E0308]
-    // note: you might have meant to use field `b` of type `B`
 }

--- a/src/test/ui/suggestions/field-access.rs
+++ b/src/test/ui/suggestions/field-access.rs
@@ -10,6 +10,11 @@ enum B {
     Snd,
 }
 
+union Foo {
+    bar: u32,
+    qux: f32,
+}
+
 fn main() {
     let a = A { b: B::Fst };
     if let B::Fst = a {}; //~ ERROR mismatched types [E0308]
@@ -19,5 +24,12 @@ fn main() {
         //~| HELP you might have meant to use field `b` of type `B`
         B::Fst => (), //~ ERROR mismatched types [E0308]
         B::Snd => (), //~ ERROR mismatched types [E0308]
+    }
+
+    let foo = Foo { bar: 42 };
+    match foo {
+        //~^ HELP you might have meant to use field `bar` of type `u32`
+        1u32 => (), //~ ERROR mismatched types [E0308]
+        _ => (),
     }
 }

--- a/src/test/ui/suggestions/field-access.stderr
+++ b/src/test/ui/suggestions/field-access.stderr
@@ -14,6 +14,39 @@ help: you might have meant to use field `b` of type `B`
 LL |     if let B::Fst = a.b {};
    |                     ^^^
 
-error: aborting due to previous error
+error[E0308]: mismatched types
+  --> $DIR/field-access.rs:16:9
+   |
+LL |     Fst,
+   |     --- unit variant defined here
+...
+LL |     match a {
+   |           - this expression has type `A`
+LL |         B::Fst => (),
+   |         ^^^^^^ expected struct `A`, found enum `B`
+   |
+help: you might have meant to use field `b` of type `B`
+   |
+LL |     match a.b {
+   |           ^^^
+
+error[E0308]: mismatched types
+  --> $DIR/field-access.rs:17:9
+   |
+LL |     Snd,
+   |     --- unit variant defined here
+...
+LL |     match a {
+   |           - this expression has type `A`
+LL |         B::Fst => (),
+LL |         B::Snd => (),
+   |         ^^^^^^ expected struct `A`, found enum `B`
+   |
+help: you might have meant to use field `b` of type `B`
+   |
+LL |     match a.b {
+   |           ^^^
+
+error: aborting due to 3 previous errors
 
 For more information about this error, try `rustc --explain E0308`.

--- a/src/test/ui/suggestions/field-access.stderr
+++ b/src/test/ui/suggestions/field-access.stderr
@@ -1,5 +1,5 @@
 error[E0308]: mismatched types
-  --> $DIR/field-access.rs:15:12
+  --> $DIR/field-access.rs:20:12
    |
 LL |     Fst,
    |     --- unit variant defined here
@@ -15,7 +15,7 @@ LL |     if let B::Fst = a.b {};
    |                     ^^^
 
 error[E0308]: mismatched types
-  --> $DIR/field-access.rs:20:9
+  --> $DIR/field-access.rs:25:9
    |
 LL |     Fst,
    |     --- unit variant defined here
@@ -32,7 +32,7 @@ LL |     match a.b {
    |           ^^^
 
 error[E0308]: mismatched types
-  --> $DIR/field-access.rs:21:9
+  --> $DIR/field-access.rs:26:9
    |
 LL |     Snd,
    |     --- unit variant defined here
@@ -48,6 +48,20 @@ help: you might have meant to use field `b` of type `B`
 LL |     match a.b {
    |           ^^^
 
-error: aborting due to 3 previous errors
+error[E0308]: mismatched types
+  --> $DIR/field-access.rs:32:9
+   |
+LL |     match foo {
+   |           --- this expression has type `Foo`
+LL |
+LL |         1u32 => (),
+   |         ^^^^ expected union `Foo`, found `u32`
+   |
+help: you might have meant to use field `bar` of type `u32`
+   |
+LL |     match unsafe { foo.bar } {
+   |           ^^^^^^^^^^^^^^^^^^
+
+error: aborting due to 4 previous errors
 
 For more information about this error, try `rustc --explain E0308`.

--- a/src/test/ui/suggestions/field-access.stderr
+++ b/src/test/ui/suggestions/field-access.stderr
@@ -1,0 +1,19 @@
+error[E0308]: mismatched types
+  --> $DIR/field-access.rs:12:12
+   |
+LL |     Fst,
+   |     --- unit variant defined here
+...
+LL |     if let B::Fst = a {};
+   |            ^^^^^^   - this expression has type `A`
+   |            |
+   |            expected struct `A`, found enum `B`
+   |
+help: you might have meant to use field `b` of type `B`
+   |
+LL |     if let B::Fst = a.b {};
+   |                     ^^^
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0308`.

--- a/src/test/ui/suggestions/field-access.stderr
+++ b/src/test/ui/suggestions/field-access.stderr
@@ -15,13 +15,14 @@ LL |     if let B::Fst = a.b {};
    |                     ^^^
 
 error[E0308]: mismatched types
-  --> $DIR/field-access.rs:19:9
+  --> $DIR/field-access.rs:20:9
    |
 LL |     Fst,
    |     --- unit variant defined here
 ...
 LL |     match a {
    |           - this expression has type `A`
+...
 LL |         B::Fst => (),
    |         ^^^^^^ expected struct `A`, found enum `B`
    |
@@ -31,14 +32,14 @@ LL |     match a.b {
    |           ^^^
 
 error[E0308]: mismatched types
-  --> $DIR/field-access.rs:20:9
+  --> $DIR/field-access.rs:21:9
    |
 LL |     Snd,
    |     --- unit variant defined here
 ...
 LL |     match a {
    |           - this expression has type `A`
-LL |         B::Fst => (),
+...
 LL |         B::Snd => (),
    |         ^^^^^^ expected struct `A`, found enum `B`
    |

--- a/src/test/ui/suggestions/field-access.stderr
+++ b/src/test/ui/suggestions/field-access.stderr
@@ -1,5 +1,5 @@
 error[E0308]: mismatched types
-  --> $DIR/field-access.rs:12:12
+  --> $DIR/field-access.rs:15:12
    |
 LL |     Fst,
    |     --- unit variant defined here
@@ -15,7 +15,7 @@ LL |     if let B::Fst = a.b {};
    |                     ^^^
 
 error[E0308]: mismatched types
-  --> $DIR/field-access.rs:16:9
+  --> $DIR/field-access.rs:19:9
    |
 LL |     Fst,
    |     --- unit variant defined here
@@ -31,7 +31,7 @@ LL |     match a.b {
    |           ^^^
 
 error[E0308]: mismatched types
-  --> $DIR/field-access.rs:17:9
+  --> $DIR/field-access.rs:20:9
    |
 LL |     Snd,
    |     --- unit variant defined here

